web3/_utils/abi.py

```diff
@@ -613,28 +573,18 @@ def __new__(cls, iterable):
         return super().__new__(cls, *iterable)
 
 
-def abi_sub_tree(data_type, data_value):
-    if data_type is None:
+def abi_sub_tree(abi_type: Optional[Union[TypeStr, ABIType]], data_value: Any) -> ABITypedData:
+    if abi_type is None:
         return ABITypedData([None, data_value])
 
-    try:
-        base, sub, arrlist = data_type
-    except ValueError:
-        base, sub, arrlist = process_type(data_type)
-
-    collapsed = collapse_type(base, sub, arrlist)
-
-    if arrlist:
-        sub_type = (base, sub, arrlist[:-1])
-        return ABITypedData([
-            collapsed,
-            [
-                abi_sub_tree(sub_type, sub_value)
-                for sub_value in data_value
-            ],
-        ])
+    if isinstance(abi_type, TypeStr):
+        abi_type = parse(abi_type)
+
+    if abi_type.is_array:
+        it = abi_type.item_type
+        return ABITypedData([str(abi_type), [abi_sub_tree(it, i) for i in data_value]])
     else:
-        return ABITypedData([collapsed, data_value])
+        return ABITypedData([str(abi_type), data_value])
```

Comment:
Something a *little* worrying about using `str` casting for this.  Maybe this deserves a formal api like `abi_type.to_canonical_type_str()`?

Reply:
@pipermerriam I'd be fine with doing that, but can you give some examples of what you think the risks are in using `str`?  In a way, I almost prefer implementing magic methods and using built-ins because then people can sort of infer how to do things without needing to look up and read about a method with a custom name.

Reply:
It's implicit functionality that calling `str(abi_type_instance)` will return a canonical ABI type string.  It'd be easy for us to change the `__str__` implementation somewhere to be more informative or provide extra metadata information which would cause this to break.

Reply:
Updated `ABIType.to_type_str` in eth-abi project and removed `__str__` API.